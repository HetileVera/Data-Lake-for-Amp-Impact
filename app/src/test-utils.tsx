import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
