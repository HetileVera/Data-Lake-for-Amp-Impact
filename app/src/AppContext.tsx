import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
