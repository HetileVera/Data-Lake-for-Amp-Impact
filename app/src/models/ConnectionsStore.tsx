import { useEffect } from "react";
import isEmpty from "lodash/isEmpty";
import { types, Instance, getParent } from "mobx-state-tree";

import { useInstallation } from "AppContext";
import { listAppflowConnections } from "api/list-appflow-connections";

import { BaseStore, isStoreNew, isStoreError, isStoreReady, isStoreLoading, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

/**
 * Represents the store for the AppFlow Salesforce connection names
 */
export const ConnectionsStore = BaseStore.named("ConnectionsStore")
  .props({
    connectionNames: types.array(types.string),
  })

  .views((self) => ({
    get empty() {
      return isEmpty(self.connectionNames);
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },
  }))

  .actions((self) => {
    const superReset = self.reset;
    return {
      reset() {
        superReset();
        self.connectionNames.clear();
      },

      async doLoad() {
        const accessKey = self.credentials.accessKey;
        const secretKey = self.credentials.secretKey;
        const region = self.region;

        const rawConnectionNames = await listAppflowConnections({ accessKey, secretKey, region });
        self.runInAction(() => {
          self.connectionNames.replace(rawConnectionNames);
        });
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IConnectionsStore extends Instance<typeof ConnectionsStore> {}

export function useConnectionsStore() {
  const installation = useInstallation();
  const store = installation.connectionsStore;

  useEffect(() => {
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
