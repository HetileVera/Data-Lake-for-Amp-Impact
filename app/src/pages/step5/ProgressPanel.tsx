import React from "react";
import { Box } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useInstallation } from "AppContext";

import { OperationProgressPanel } from "pages/step5/OperationProgressPanel";

export const ProgressPanel = observer(() => {
  const installation = useInstallation();
  const deployment = installation.deploymentOperations;

  return (
    <Box p={3} pb={4}>
      {deployment.operations.map((operation, index) => (
        <OperationProgressPanel key={index} operation={operation} />
      ))}
      <Box mt="60px">
        <OperationProgressPanel operation={deployment} message="Overall Progress" />
      </Box>
    </Box>
  );
});
