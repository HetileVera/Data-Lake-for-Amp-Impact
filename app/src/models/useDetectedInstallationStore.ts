import { isStoreError, isStoreLoading, isStoreReady, isStoreReLoading } from "models/BaseStore";

import { DetectedInstallationStore } from "./DetectedInstallationsStore";

// Create an instance of this store
const store = DetectedInstallationStore.create({});

export function useDetectedInstallationStore() {
  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
