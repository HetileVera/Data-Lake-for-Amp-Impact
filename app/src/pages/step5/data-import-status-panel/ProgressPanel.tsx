import React from "react";
import { observer } from "mobx-react";
import { Progress } from "@chakra-ui/react";

import { useImportStatusStore } from "models/ImportStatusStore";
import { isStoreLoading, isStoreReLoading } from "models/BaseStore";

export const ProgressPanel = observer(({ colorScheme }: { colorScheme: string }) => {
  const { isError, store } = useImportStatusStore();

  if (isError) return null;
  if (!isStoreLoading(store) && !isStoreReLoading(store)) return null;

  return <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={`${colorScheme}.100`} borderRadius="lg" mb={3} />;
});
