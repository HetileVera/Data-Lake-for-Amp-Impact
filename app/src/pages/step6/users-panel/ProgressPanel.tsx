import React from "react";
import { Progress } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useInstallation } from "AppContext";
import { useColorScheme } from "models/useColorScheme";
import { useUsersStore } from "models/useUsersStore";

export const ProgressPanel = observer(() => {
  const installation = useInstallation();
  const { isError, isLoading, isReloading } = useUsersStore(installation);
  const { tone, colorScheme } = useColorScheme();

  if (isError) return null;
  if (!isLoading && !isReloading) return null;

  return <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={tone(100)} borderRadius="lg" mb={4} mt={0} />;
});
