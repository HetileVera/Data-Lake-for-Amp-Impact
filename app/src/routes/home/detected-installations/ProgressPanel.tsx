import React from "react";
import { Progress, Text } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useDetectedInstallationStore } from "models/useDetectedInstallationStore";
import { useColorScheme } from "models/useColorScheme";

export const ProgressPanel = observer(() => {
  const { isError, isLoading, isReloading } = useDetectedInstallationStore();
  const { tone, colorScheme } = useColorScheme();

  if (isError) return null;
  if (!isLoading && !isReloading) return null;

  return (
    <>
      <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={tone(100)} borderRadius="lg" mb={1} />
      <Text display="block" textAlign="center" color={tone(800)} fontSize="sm" mb={6}>
        Processing
      </Text>
    </>
  );
});
