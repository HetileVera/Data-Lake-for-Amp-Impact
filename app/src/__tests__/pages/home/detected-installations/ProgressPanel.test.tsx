import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, CustomChakraProvider } from "test-utils";

import { StoreProvider } from "AppContext";

import { ProgressPanel } from "pages/home/detected-installations/ProgressPanel";

import * as detectedInstallationStore from "models/useDetectedInstallationStore";

describe("DetectedInstallations -> ProgressPanel component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    const tree = create(
      <CustomChakraProvider>
        <StoreProvider>
          <ProgressPanel />
        </StoreProvider>
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly with an error", () => {
    // @ts-ignore
    jest.spyOn(detectedInstallationStore, "useDetectedInstallationStore").mockImplementation(() => {
      return {
        isError: true,
        isLoading: false,
        isReloading: false,
      };
    });

    render(
      <StoreProvider>
        <ProgressPanel />
      </StoreProvider>
    );

    expect(screen.findByRole("body")).toBeTruthy();
    expect(screen.queryByText(/processing/i)).not.toBeInTheDocument();
  });

  test("render properly with not loading and not reloading", () => {
    // @ts-ignore
    jest.spyOn(detectedInstallationStore, "useDetectedInstallationStore").mockImplementation(() => {
      return {
        isError: false,
        isLoading: false,
        isReloading: false,
      };
    });

    render(
      <StoreProvider>
        <ProgressPanel />
      </StoreProvider>
    );

    expect(screen.findByRole("body")).toBeTruthy();
    expect(screen.queryByText(/processing/i)).not.toBeInTheDocument();
  });

  test("render properly with processing when loading", () => {
    // @ts-ignore
    jest.spyOn(detectedInstallationStore, "useDetectedInstallationStore").mockImplementation(() => {
      return {
        isError: false,
        isLoading: true,
        isReloading: false,
      };
    });

    render(
      <StoreProvider>
        <ProgressPanel />
      </StoreProvider>
    );

    expect(screen.queryByText(/processing/i)).toBeInTheDocument();
  });

  test("render properly with processing when reloading", () => {
    // @ts-ignore
    jest.spyOn(detectedInstallationStore, "useDetectedInstallationStore").mockImplementation(() => {
      return {
        isError: false,
        isLoading: false,
        isReloading: true,
      };
    });

    render(
      <StoreProvider>
        <ProgressPanel />
      </StoreProvider>
    );

    expect(screen.queryByText(/processing/i)).toBeInTheDocument();
  });
});
