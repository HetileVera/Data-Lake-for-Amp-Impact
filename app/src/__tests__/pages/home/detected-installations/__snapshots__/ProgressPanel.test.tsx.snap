// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetectedInstallations -> ProgressPanel component snapshot 1`] = `
Array [
  <div
    data-testid="test-root"
  />,
  <span />,
]
`;
